src/Evolution/Systems/ScalarWave/Equations.cpp

```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
     dt_phi->get(d) = -d_pi.get(d);
   }
 }
+/// \endcond
+
+template <size_t Dim>
+void ComputeNormalDotFluxes<Dim>::apply(
+    const gsl::not_null<Scalar<DataVector>*> pi_normal_dot_flux,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        phi_normal_dot_flux,
+    const gsl::not_null<Scalar<DataVector>*> psi_normal_dot_flux,
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        interface_unit_normal) noexcept {
+  // We assume that all values of psi_normal_dot_flux are the same. The reason
+  // is that std::fill is actually surprisingly/disappointingly slow.
+  if (psi_normal_dot_flux->get()[0] != 0.0) {
+    std::fill(psi_normal_dot_flux->get().begin(),
+              psi_normal_dot_flux->get().end(), 0.0);
+  }
+
+  get(*pi_normal_dot_flux) = get<0>(interface_unit_normal) * get<0>(phi);
+  for (size_t i = 1; i < Dim; ++i) {
+    get(*pi_normal_dot_flux) += interface_unit_normal.get(i) * phi.get(i);
+  }
+
+  for (size_t i = 0; i < Dim; ++i) {
+    phi_normal_dot_flux->get(i) = interface_unit_normal.get(i) * get(pi);
+  }
+}
+
+template <size_t Dim>
+void UpwindFlux<Dim>::package_data(
+    const gsl::not_null<Variables<package_tags>*> packaged_data,
+    const Scalar<DataVector>& normal_dot_flux_pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& normal_dot_flux_phi,
+    const Scalar<DataVector>& /*normal_dot_flux_psi*/,
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal)
+    const noexcept {
+  // Computes the contribution to the numerical flux from one side of the
+  // interface.
+  //
+  // The packaged_data stores:
+  // <Pi> = 0.5 * pi
+  // <NormalDotFlux<Pi>> = 0.5 * normal_dot_flux_pi
+  // <NormalTimesFluxPi_i> = 0.5 * normal_dot_flux_pi * n_i
+  // <NormalDotFlux<Phi>_i> = 0.5 * normal_dot_flux_phi_i
+  //
+  // Note: There is a minus sign present now when sending to the neighbor
+  // because the normal is my normal, but I used the one pointing in the
+  // opposite direction to the one my neighbor chooses.
+  get(get<Pi>(*packaged_data)) = 0.5 * get(pi);
+  get<Tags::NormalDotFlux<Pi>>(*packaged_data) =
+      Scalar<DataVector>(0.5 * get(normal_dot_flux_pi));
+  auto& normal_times_flux_pi = get<NormalTimesFluxPi>(*packaged_data);
+  for (size_t d = 0; d < Dim; ++d) {
+    normal_times_flux_pi.get(d) =
+        0.5 * interface_unit_normal.get(d) * get(normal_dot_flux_pi);
+    get<Tags::NormalDotFlux<Phi<Dim>>>(*packaged_data).get(d) =
+        0.5 * normal_dot_flux_phi.get(d);
+  }
+}
+
+template <size_t Dim>
+void UpwindFlux<Dim>::operator()(
```

Comment:
okay, I think to make this absolutely clear, the input arguments to this function should be named what they are: